swift/ql/lib/upgrades/147e087e57e51b2eb41e75c9c97380d0e6c20ecb/for_in_upgrade.ql

```diff
@@ -0,0 +1,171 @@
+  pattern_binding_decl_inits(id, index, init)
+  or
+  exists(ForEachStmt foreach |
+    Fresh::map(TIteratorVar(foreach)) = id and
+    foreach.getSequence() = init and
+    index = 0
+  )
+}
+
+query predicate for_each_stmt_next_calls(ForEachStmt id, NewElement nextCall) {
+  Fresh::map(TNextCall(id)) = nextCall
+}
+
+query predicate new_dot_syntax_call_exprs(NewElement id) {
+  dot_syntax_call_exprs(id)
+  or
+  Fresh::map(TNextCallMethodLookup(_)) = id
+}
+
+query predicate new_self_apply_exprs(NewElement id, NewElement base) {
+  self_apply_exprs(id, base)
+  or
+  exists(ForEachStmt foreach |
+    Fresh::map(TNextCallMethodLookup(foreach)) = id and
+    Fresh::map(TNextCallInOutConversion(foreach)) = base
+  )
+}
+
+query predicate new_in_out_exprs(NewElement inOutExpr, NewElement subExpr) {
+  in_out_exprs(inOutExpr, subExpr)
+  or
+  exists(ForEachStmt foreach |
+    Fresh::map(TNextCallInOutConversion(foreach)) = inOutExpr and
+    Fresh::map(TNextCallVarRef(foreach)) = subExpr
+  )
+}
+
+query predicate new_apply_exprs(NewElement id, NewElement func) {
+  apply_exprs(id, func)
+  or
+  exists(ForEachStmt foreach |
+    Fresh::map(TNextCall(foreach)) = id and
+    Fresh::map(TNextCallMethodLookup(foreach)) = func
+  )
+  or
+  exists(ForEachStmt foreach |
+    Fresh::map(TNextCallMethodLookup(foreach)) = id and
+    Fresh::map(TNextCallFuncRef(foreach)) = func
+  )
+}
+
+Element getParent(Element type) {
+  any_generic_type_parents(type, result)
+  or
+  // there may be an extension that implements IteratorProtocol
+  exists(@extension_decl extDecl, @nominal_type_decl typeDecl, @protocol_decl protocolDecl |
+    any_generic_types(type, typeDecl) and
+    extension_decls(extDecl, typeDecl) and
+    extension_decl_protocols(extDecl, _, protocolDecl) and
+    any_generic_types(result, protocolDecl)
+  )
+}
+
+// TODO: do we need a new_apply_expr_arguments
+query predicate new_decl_ref_exprs(NewElement id, NewElement decl) {
+  decl_ref_exprs(id, decl)
+  or
+  exists(ForEachStmt foreach |
+    Fresh::map(TNextCallVarRef(foreach)) = id and
+    Fresh::map(TIteratorVarPattern(foreach)) = decl
+  )
+  or
+  exists(ForEachStmt foreach, @element sequence, @element exprType, @element parentType, @element typeDecl |
+    Fresh::map(TNextCallFuncRef(foreach)) = id and
+    // ForEachStmt.getSequence().getType().getMember(next)
+    sequence = foreach.getSequence() and
+    expr_types(sequence, exprType) and
+    parentType = getParent*(exprType) and
+    any_generic_types(parentType, typeDecl) and
+    decl_members(typeDecl, _, decl) and
+    callable_names(decl, "next()")
+  )
+}
+
+query predicate new_lookup_exprs(NewElement id, NewElement base) {
+  lookup_exprs(id, base)
+  or
+  exists(ForEachStmt foreach |
+    Fresh::map(TNextCallMethodLookup(foreach)) = id and
+    Fresh::map(TNextCallVarRef(foreach)) = base
+  )
+}
+
+query predicate new_call_exprs(NewElement id) {
+  call_exprs(id) or
+  Fresh::map(TNextCall(_)) = id
+}
```

Comment:
The upgrade script looks plausible to me.  My understanding is that we edit the `for_each_stmts` relation, and synthesize a bunch of relations under it that would be extracted correctly in the new version and that we now depend on.  It looks something like this:
```
for_each_stmts
	pattern
	sequence --- removed in the for_each_stmts
	where
	iteratorVar --- added in pattern_binding_decls, linked with for_each_stmt_iterator_vars
		iteratorVarPattern --- added in named_patterns, linked with pattern_binding_decl_patterns
		init --- added in pattern_binding_decl_inits with the old 'sequence' Expr
	nextCall --- added in call_exprs, linked with for_each_stmt_next_calls
		nextCallMethodLookup --- added in dot_syntax_call_exprs, linked in apply_exprs
			nextCallInOutConversion --- linked in self_apply_exprs
				nextCallVarRef --- linked in in_out_exprs
					nextCallVarRefPattern --- linked in decl_ref_exprs, lookup_exprs
			nextCallFuncRef --- linked in apply_exprs, decl_ref_exprs
	body
```
This is fiddly stuff.  What testing have you done / do you plan to do?

I assume the downgrade script is coming as well?  Do we need to remove all this structure in the downgrade or will it be sufficient to simply unlink it???

Reply:
For testing so far I've been running the AST, CFG, and data flow qltests with the old extractor and the new library. There's still a couple things that aren't working, and I think I'll need to add types for all the expressions, but I wanted to get something uploaded.

I think we'll need to remove all the structure in the downgrade script. Otherwise we end up with floating entities that could mess with other queries